kafkaesk/ext/logging/handler.py

```diff
@@ -0,0 +1,111 @@
+from .formatter import PydanticLogModel
+from .record import PydanticLogRecord
+from pydantic import BaseModel
+from typing import IO
+from typing import Optional
+from typing import Tuple
+
+import asyncio
+import kafkaesk
+import logging
+import sys
+
+
+class InvalidLogFormat(Exception):
+    ...
+
+
+class PydanticStreamHandler(logging.StreamHandler):
+    def __init__(self, stream: Optional[IO[str]] = None):
+        super().__init__(stream=stream)
+
+    def format(self, record: logging.LogRecord) -> str:
+        message = super().format(record)
+
+        if isinstance(message, BaseModel):
+            message = message.json()
+
+        return message
+
+
+class LogQueue:
+    def __init__(self, app: kafkaesk.Application):
+        self._queue: asyncio.Queue[Tuple[str, BaseModel]] = asyncio.Queue()
+        self._app = app
+
+    async def run(self) -> None:
+        while True:
+            stream, message = await self._queue.get()
+            try:
+                await self._app.publish(stream, message)
+            except Exception as err:
```

Comment:
Having the feeling that `asyncio.CancelledError` would not need to be masked and in case of having this exception just to stop the iteration and finish the `run` execution

Reply:
@vangheem @pfreixes 

I've got a couple of questions around the actual integration of the Kafkaesk application:

1. What is the expected behavior if a log message fails to publish?  My initial thought is to re-queue the message and attempt to send it again later.  If we attempt to re-queue should we have a limit to the number of re-queue attempts?  If we ultimately fail to send a message should we silently drop the message, log the message to standard out, or raise an exception and let the logging framework handle it?

2. How should we expect the logger the behave if the Kafkaesk application is closed / stopped?  Currently the `publish` function will attempt to re-initialize the application, meaning the application could be closed / finalized but any log messages would trigger a re-initialization

Reply:
My other option here is to not accept a kafkesk application and construct my own application within the handler / queue.  I then don't need to worry about the application being updated anywhere outside of the logger code.

Reply:
1) IMO, log messages should be non-essential logging of any application. Failures should be outputted to sys.stderr and dropped. If you have essential data to log, you should use the standard publish and commit pattern

2) What do you think about requiring providing an application to use with it? If it is closed while trying to flush logs, throw errors to stderr(do not try to reinit).

Also, could we register log handlers cleanup callbacks on the application instead so when the application instance is in teardown mode(finalize), it can attempt to wait for flushing of remaining queue?

Reply:
>What do you think about requiring providing an application to use with it? If it is closed while trying to flush logs, throw errors to stderr(do not try to reinit).

Would this be a `Kafkaesk.app.Application` class you mean to provide?  That happens today and adding a check to ensure the application is initialized would be fairly trivial.

>Also, could we register log handlers cleanup callbacks on the application instead so when the application instance is in teardown mode(finalize), it can attempt to wait for flushing of remaining queue?

This would be updating `Kafkaesk.app.Application.publish()` to execute the callbacks prior to tearing down it's producers correct?  That's a pretty simple way to approach the issue, I like it.

Reply:
I was actually thinking it should be just on setup.

I think instead it would be something like this in your handler init:

```
def __init__(self, app):
    ...
    app.on('finialize', self.finalize)

async def finalize(self):
    await queue.join() # or some other way to flush
```

Then in Application:

```
def __init__(self):
    self._event_handlers = {}

def on(self, name, handler):
    if name not in self._event_handlers:
        self._event_handlers[name] = []
    self._event_handler[name].append(handler)

def finalize(self):
    ...
    for ename, handlers in self._event_handler.items():
        for handler in handlers:
            await handler()

```